src/box/txn_limbo.c

```diff
@@ -1163,6 +1227,10 @@ txn_limbo_req_prepare(struct txn_limbo *limbo,
 	case IPROTO_RAFT_PROMOTE:
 	case IPROTO_RAFT_DEMOTE: {
 		assert(limbo->svp_confirmed_lsn == -1);
+		/*
+		 * Do not update the `volatile_confirmed_lsn` here, this is
+		 * handled by `limbo->is_in_rollback`.
+		 */
```

Comment:
Hm? Does it mean in the other places the volatile LSN can go down when you assign it? I have a feeling we are going to get nasty bugs when we allow that LSN to go down in any random place and not just in one single place like here.

Reply:
Shouldn't we update `volatile_confirmed_vclock` here and then update `confirmed_vclock` in the `read_promote/demote`?

Reply:
As I see it, the worker fiber should not interfere when the `limbo->is_in_rollback`. 

> Hm? Does it mean in the other places the volatile LSN can go down when you assign it? I have a feeling we are going to get nasty bugs when we allow that LSN to go down in any random place and not just in one single place like here.

I didn't get your point. If `limbo->is_in_rollback`, it does not accept new transactions, and we make sure the confirm is not written concurrently.

Technically, `volatile_confirmed_lsn` can go down after reading promote/demote, just like `confirmed_lsn` can, since we will start looking at another instance's transactions.

LSN cannot go down in any random place, this patch does not change the logic of promote/demote.

> Shouldn't we update `volatile_confirmed_vclock` here and then update `confirmed_vclock` in the read_promote/demote?

What's the point of introducing a `volatile_confirmed_vclock`, if we only want to asynchronously write confirms? I.e., the LSNs of transactions coming from a fixed instance.